_posts/2019-10-18-vue-3-composition-api.md

```diff
@@ -0,0 +1,190 @@
+
+```javascript
+<template>
+  <button @click="increment">
+    Count is: {{ state.count }}, double is: {{ state.double }}
+  </button>
+</template>
+
+<script>
+import { reactive, computed } from 'vue'
+
+export default {
+  setup() {
+    const state = reactive({
+      count: 0,
+      double: computed(() => state.count * 2)
+    })
+
+    function increment() {
+      state.count++
+    }
+
+    return {
+      state,
+      increment
+    }
+  }
+}
+</script>
+```
+
+## Setup Function
+
+The setup function is where you put your data and functions for your feature. You could also move your feature code into its own function outside of the setup or its own javascript file and then would have to declare the code in the setup function and return it to use.
+
+- Takes two optional arguments props, and context
+- Create methods, data state, functions inside setup function
+- All of the methods, functions need to be returned from the setup function
+- Shared code can be imported and then declared in the setup function (must be returned)
+- Does not have access to this
+
+## Refs vs Reactive
+
+There are two ways to deal with state and reactivity in the setup function, _refs_ and _reactive_. Using one over the other is a matter of preference and coding style. They both allow Vue to keep track of your state.
+
+- Need to be imported to be used
+- Data using _refs_ or _reactive_ need to be returned as objects from setup function
+
+Knowing when to use refs and reactive can be confusing and best practices are still being [developed](https://vue-composition-api-rfc.netlify.com/#ref-vs-reactive)
+
+- Refs are great for declaring single variables
+- Reactive is great for listing all your state inside objects
+- Reactive objects cannot be destructured or spread without using the _toRefs method_
+
+## Working with refs
+
+[Ref example](https://vue-composition-api-rfc.netlify.com/#api-introduction)
+
+```javascript
+<template>
+  <button @click="increment">
+    Count is: {{ count.value }}, double is: {{ count.value * 2 }}
+  </button>
+</template>
+
+<script>
+  import { ref, watch } from 'vue'
+
+  let count = ref(0)
+
+  function increment() {
+    count.value++
+  }
+
+  const renderContext = {
+    count,
+    increment
+  }
+
+  watch(() => {
+    renderTemplate(
+      `<button @click="increment">{{ count }}</button>`,
+      renderContext
+    )
+  })
+</script>
+```
+
+Refs in the setup function work differently then refs in the rest of Vue. In the setup function when using a ref you declare the variable and set the initial value using `ref()`.
+{% raw %}`let todo = ref('')`{% endraw %}
+Working with your data in the template section of the component you need to reference the data with the name of the ref and _add .value_ at the end
+
+{% raw %} `<h1> {{ todo.value }} </h1>` {% endraw %}
+
+## Working with reactive
+
+Reactive takes an object and returns a reactive object
+
+```javascript
+setup() {
+Const todos = reactive({
+item:’’
```

Comment:
Is this the intent?

```suggestion
item: ''
```

This code block could be cleaned up a bit formatting wise.

Reply:
yes thanks for catching this. I wrote this post initially in google sheets and I think it messed up the quotes